library/Icingadb/Model/CustomvarFlat.php

```diff
@@ -130,7 +134,33 @@ public function unFlattenVars(Traversable $flattenedVars): array
                 $source = [$realName => $var->customvar->value];
 
                 $sourcePath = ltrim(substr($var->flatname, strlen($realName)), '.');
-                $path = array_merge([$realName], $sourcePath ? explode('.', $sourcePath) : []);
+                $sourcePaths = [];
+
+                // This will split the source path and compute each time from right to left.
+                // Example: begin -> `foo[0][2][]` -> `foo[0][2]` -> `foo[0]` -> `foo` this will end the loop.
+                while (preg_match('/^\w+(\[(\d*)])+$/', $sourcePath, $m)) {
+                    if ($m[2] !== "") {
+                        // When it's not an empty array list e.g `foo[]`
+                        $sourcePaths[] = $m[2];
+                    }
+
+                    $sourcePath = substr($sourcePath, 0, -strlen($m[1]));
+                }
+
+                if (! empty($sourcePaths)) {
+                    // Append the last remaining custom var flat name. From the above example the remaining custom var
+                    // path is just `foo` and we will reverse these paths when joining below.
+                    $sourcePaths[] = $sourcePath;
+
+                    // Join the custom var paths with the real name in reverse order.
+                    $path = array_merge([$realName], array_reverse($sourcePaths));
+                } else {
+                    if ($sourcePath === '[]') {
+                        $sourcePath = null;
+                    }
+
+                    $path = array_merge([$realName], $sourcePath ? explode('.', $sourcePath) : []);
+                }
```

Comment:
Should suffice, please test:

```suggestion
                $path = array_merge([$realName], $sourcePath ? preg_split('/(?<=\w)\.|(?=\[)/', $sourcePath) : []);
```

Reply:
**Without your changes:**
<img width="303" alt="Bildschirmfoto 2023-06-23 um 15 34 16" src="https://github.com/Icinga/icingadb-web/assets/57616252/c8d94893-3914-4e3e-8a61-b8436c26704a">

**With your changes:**
<img width="303" alt="Bildschirmfoto 2023-06-23 um 15 33 30" src="https://github.com/Icinga/icingadb-web/assets/57616252/62d5a534-7b02-4667-8d21-3e63e3cdd368">

Reply:
```suggestion
                $path = array_merge([$realName], $sourcePath ? preg_split('(?<=\w|\])\.|(?=\[)', $sourcePath) : []);
```